### PR TITLE
Koa 5877 implement custom ksp processor for stories

### PR DESCRIPTION
**Pull request by bvitaliyg**
+ [ ] Check this if you have read and followed the [contributing guidelines](https://github.com/Skyscanner/backpack-android/blob/main/CONTRIBUTING.md)

Remember to include the following changes:
+ [ ] Component `README.md`
+ [ ] Tests

If you are curious about how we review, please read through the [code review guidelines](https://github.com/Skyscanner/backpack/blob/main/CODE_REVIEW_GUIDELINES.md)

### CODE REVIEW

File: buildSrc/build.gradle.kts [hunk 1]
```diff
@@ -16,7 +16,7 @@ dependencies {
   implementation("com.google.guava:guava:31.1-jre")
   implementation("com.google.code.gson:gson:2.10.1")
 
-  implementation(platform("org.http4k:http4k-bom:4.39.0.0"))
+  implementation(platform("org.http4k:http4k-bom:4.36.0.0"))
```
Comment (bvitaliyg) [inline buildSrc/build.gradle.kts:19]: This causes compilating error in my case

File: ksp/src/main/kotlin/net/skyscanner/backpack/ksp/visitor/ViewStoriesVisitor.kt [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import com.google.devtools.ksp.symbol.KSName
+import com.google.devtools.ksp.symbol.KSNode
+import com.google.devtools.ksp.visitor.KSDefaultVisitor
+import net.skyscanner.backpack.ksp.ComponentDefinition
+import net.skyscanner.backpack.ksp.StoryDefinition
+import net.skyscanner.backpack.ksp.ViewStoryAnnotation
+import net.skyscanner.backpack.ksp.find
+import net.skyscanner.backpack.ksp.get
+
+object ViewStoriesVisitor : KSDefaultVisitor<Map<KSName, ComponentDefinition>, StoryDefinition?>() {
```
Comment (marianeum) [inline ksp/src/main/kotlin/net/skyscanner/backpack/ksp/visitor/ViewStoriesVisitor.kt:32]: looks like the view/compose classes are almost the same - can some bits be extracted?

File: ksp/src/main/kotlin/net/skyscanner/backpack/ksp/visitor/ComposeStoriesVisitor.kt [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    val componentAnnotation = function.annotations.firstNotNullOfOrNull {
+      it.annotationType.resolve().declaration.qualifiedName?.let(data::get)
+    }
+
+    return when {
+      annotation != null && location is FileLocation -> StoryDefinition(
+        component = componentAnnotation ?: error("No component definition is found!"),
+        name = annotation[ComposeStoryAnnotation.paramName],
+        isCompose = true,
+        isScreenshot = annotation[ComposeStoryAnnotation.paramScreenshot],
```
Comment (marianeum) [inline ksp/src/main/kotlin/net/skyscanner/backpack/ksp/visitor/ComposeStoriesVisitor.kt:47]: we're not supporting screenshot bits yet, right?

Comment (bvitaliyg) [reply ksp/src/main/kotlin/net/skyscanner/backpack/ksp/visitor/ComposeStoriesVisitor.kt:47]: It's just easier to have it there, and transferring from spike branch was also easier

File: ksp/src/main/kotlin/net/skyscanner/backpack/ksp/visitor/ViewStoriesVisitor.kt (hunk 2 above)
Comment (bvitaliyg) [reply ksp/src/main/kotlin/net/skyscanner/backpack/ksp/visitor/ViewStoriesVisitor.kt:32]: I was thinking about that, but it'll make the whole stuff more complicated to read and maintain. And it's just 2 classes

File: app/src/main/java/net/skyscanner/backpack/demo/ui/CasesScreen.kt [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+import net.skyscanner.backpack.compose.tokens.Settings
+import net.skyscanner.backpack.demo.R
+import net.skyscanner.backpack.demo.SettingsActivity
+import net.skyscanner.backpack.demo.compose.ComponentItem
+import net.skyscanner.backpack.demo.meta.Component
+import net.skyscanner.backpack.demo.meta.Story
+import net.skyscanner.backpack.demo.meta.all
+
+@Composable
+fun CasesScreen(
```
[the remark this answers is not in the document]
Comment (bvitaliyg) [reply app/src/main/java/net/skyscanner/backpack/demo/ui/CasesScreen.kt:48]: It shows not only tabs but lists of cases as well. Haven't found better naming :)

File: app/src/main/java/net/skyscanner/backpack/demo/ui/ComponentsScreen.kt [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+import net.skyscanner.backpack.demo.compose.ComponentItem
+import net.skyscanner.backpack.demo.compose.ComponentsTitle
+import net.skyscanner.backpack.demo.meta.Component
+import net.skyscanner.backpack.demo.meta.Story
+import net.skyscanner.backpack.demo.meta.all
+
+@Composable
+fun ComponentsScreen(
+  modifier: Modifier = Modifier,
+  stories: List<Story> = remember { Story.all() },
```
[the remark this answers is not in the document]
Comment (bvitaliyg) [reply app/src/main/java/net/skyscanner/backpack/demo/ui/ComponentsScreen.kt:48]: No, each component can have multiple stories, each of the stories can be view or compose

File: buildSrc/build.gradle.kts (hunk 1 above)
Comment (marianeum) [reply buildSrc/build.gradle.kts:19]: is your kotlin version 1.8 in AS? it should really use the latest version as our kotlin version is 1.8

File: app/src/main/java/net/skyscanner/backpack/demo/meta/ViewStory.kt [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+import androidx.annotation.LayoutRes
+import androidx.compose.ui.tooling.preview.Preview
+import javax.annotation.concurrent.Immutable
+
+@Immutable
+@Suppress("Detekt.PreviewNaming")
+@Target(AnnotationTarget.FUNCTION)
+@Retention(AnnotationRetention.BINARY)
+
+@Preview(name = "Default",
```
[the remark this answers is not in the document]
Comment (bvitaliyg) [reply app/src/main/java/net/skyscanner/backpack/demo/meta/ViewStory.kt:32]: We already using multiple preview annotations. It's just these annotations themselves are duplicated for View and Compose.

We can create something like DmPreview, RtlPrevew, etc, but I'd rather not create new entities – these annotations can be used for any other composable. A solution would be to extract to to a different module and make internal, but it's a bit overkill

File: app/src/main/java/net/skyscanner/backpack/demo/ui/ComponentsScreen.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (marianeum) [reply app/src/main/java/net/skyscanner/backpack/demo/ui/ComponentsScreen.kt:48]: ah i misunderstood what this one's about. how does the component/token categorisation work?

File: app/src/main/java/net/skyscanner/backpack/demo/meta/ViewStory.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (marianeum) [reply app/src/main/java/net/skyscanner/backpack/demo/meta/ViewStory.kt:32]: yeah can we create a shared definition for all three and then reuse it across view + compose. this is also something that we may want to have in backpack, rather than just here, so it can be utilised in the app as well.

also can we use the canvas colours instead of just black/white?

File: ksp/src/main/kotlin/net/skyscanner/backpack/ksp/visitor/ViewStoriesVisitor.kt (hunk 2 above)
Comment (marianeum) [reply ksp/src/main/kotlin/net/skyscanner/backpack/ksp/visitor/ViewStoriesVisitor.kt:32]: the two need to be in sync, though, so this feels harder to maintain?

File: app/build.gradle [hunk 7]
```diff
@@ -117,4 +124,5 @@ dependencies {
   androidTestImplementation rootProject.ext.facebookSnapshotTestingCore
   implementation project(':Backpack')
   implementation project(':backpack-compose')
+  ksp project(":ksp")
```
Comment (marianeum) [inline app/build.gradle:127]: can we give the module a more descriptive name? `ksp` is more the implementation detail, rather than what it actually does and if we ever need another `ksp` implementation we'll get a clash

File: ksp/src/main/kotlin/net/skyscanner/backpack/ksp/BackpackSymbolProcessor.kt [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+import com.google.devtools.ksp.processing.SymbolProcessorProvider
+import com.google.devtools.ksp.symbol.KSAnnotated
+import com.google.devtools.ksp.validate
+import net.skyscanner.backpack.ksp.visitor.ComponentsVisitor
+import net.skyscanner.backpack.ksp.visitor.ComposeStoriesVisitor
+import net.skyscanner.backpack.ksp.visitor.ViewStoriesVisitor
+import net.skyscanner.backpack.ksp.writer.writeListOfStories
+
+@OptIn(ExperimentalProcessingApi::class)
+class BackpackSymbolProcessor(
```
Comment (marianeum) [inline ksp/src/main/kotlin/net/skyscanner/backpack/ksp/BackpackSymbolProcessor.kt:35]: can you add some tests for these? as the ksp files will end up in the build folder we won't have a manual check to see if anything changed, but need to rely on tests to ensure everything is in the right format. particularly with different edge cases, view/compose versions etc

File: .github/dependabot.yml [hunk 9]
```diff
@@ -12,6 +12,12 @@ updates:
     interval: daily
     time: "10:00"
   open-pull-requests-limit: 10
+- package-ecosystem: gradle
+  directory: "/ksp"
```
Comment (marianeum) [inline .github/dependabot.yml:16]: can you update this directory please?

File: app/src/main/java/net/skyscanner/backpack/demo/ui/CasesScreen.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (marianeum) [reply app/src/main/java/net/skyscanner/backpack/demo/ui/CasesScreen.kt:48]: I'd say ultimately this is the `ComponentScreen` - `Cases` is very unclear what that means. particularly as it could just be the component details directly

File: app/src/main/java/net/skyscanner/backpack/demo/ui/ComponentsScreen.kt [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+import net.skyscanner.backpack.demo.R
+import net.skyscanner.backpack.demo.SettingsActivity
+import net.skyscanner.backpack.demo.compose.ComponentItem
+import net.skyscanner.backpack.demo.compose.ComponentsTitle
+import net.skyscanner.backpack.demo.meta.Component
+import net.skyscanner.backpack.demo.meta.Story
+import net.skyscanner.backpack.demo.meta.all
+
+@Composable
+fun ComponentsScreen(
```
Comment (marianeum) [inline app/src/main/java/net/skyscanner/backpack/demo/ui/ComponentsScreen.kt:46]: maybe `ComponentListScreen`?

File: app/src/main/java/net/skyscanner/backpack/demo/ui/DemoScreen.kt [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+import net.skyscanner.backpack.compose.navigationbar.BpkTopNavBar
+import net.skyscanner.backpack.compose.navigationbar.IconAction
+import net.skyscanner.backpack.compose.navigationbar.NavIcon
+import net.skyscanner.backpack.compose.tokens.Settings
+import net.skyscanner.backpack.demo.R
+import net.skyscanner.backpack.demo.SettingsActivity
+import net.skyscanner.backpack.demo.meta.Story
+
+@Composable
+fun DemoScreen(
```
Comment (marianeum) [inline app/src/main/java/net/skyscanner/backpack/demo/ui/DemoScreen.kt:37]: `StoryScreen`?

File: backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/annotation/BpkPreviews.kt [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+  locale = "en",
+  showBackground = true,
+  backgroundColor = 0xFFFFFFFF,
+)
+
+@Preview(name = "DM",
+  uiMode = Configuration.UI_MODE_NIGHT_YES,
+  locale = "en",
+  showBackground = true,
+  backgroundColor = 0xFF010913,
```
Comment (marianeum) [inline backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/annotation/BpkPreviews.kt:42]: can you add a comment to indicate that this is `canvas`?

File: meta-processor/build.gradle [hunk 13]
```diff
@@ -0,0 +1,12 @@
+apply plugin: 'kotlin'
+
+dependencies {
+  implementation "com.squareup:kotlinpoet:1.12.0"
```
Comment (marianeum) [inline meta-processor/build.gradle:4]: can you move the dependencies to the dependency definition to keep them in sync? that way we don't actually have to include this module in the dependabot config

**Review by marianeum — APPROVED** [review]
